### PR TITLE
Add logging

### LINKED ISSUE #144: Logging
We should improve (start) logging to make discovering the source of an error easier, especially in combination with error reporting (#37)

- [x] (Use a logging package)
- [x] Failed requests
- [x] Navigation between screens
- [x] Login/logout

### PR DESCRIPTION
**Pull request by JonasWanke**
**Fixes: #144**

- You can use `logger.x(<message>, [<error>, <stackTrace>])` (x ∈ {`v`, `d`, `i`, `w`, `e`, `wtf`}) to produce a colorful output including emojis and the time!
- Shake your device (in debug mode) to open an on-device console!

We currently log:
- login & logout
- failed network requests
- navigation events (suboptimal without route names, see 2<sup>nd</sup> screenshot)

| Failed network request | On-device console |
|:---:|:---:|
| image | image |

### CODE REVIEW

**Comment by janrenz** [comment]
should wer make a toggle for this in the setting?

**Comment by JonasWanke** [comment]
@janrenz What du you want to toggle? (Both logging in general and the on-device console are only enabled in debug builds)

**Comment by janrenz** [comment]
@JonasWanke Ah, if its only in debug lets go ..

File: lib/app/logger.dart [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+  static const middleCorner = ' ├─';
+  static const bottomCorner = ' └─';
+
+  @override
+  void log(LogEvent event) {
+    final messageStr = stringifyMessage(event.message) ?? '';
+    final timeStr = timeFormat.format(Instant.now().inLocalZone().clockTime);
+    String stackTraceStr;
+    if (event.stackTrace != null ||
+        [Level.error, Level.wtf, Level.warning].contains(event.level)) {
```
Comment (MarcelGarus) [inline lib/app/logger.dart]: We should probably list these error levels in order of their severity:
```suggestion
        [Level.error, Level.wtf].contains(event.level)) {
```

File: lib/app/services/navigator_observer.dart [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+  @override
+  void didReplace({Route<dynamic> newRoute, Route<dynamic> oldRoute}) {
+    logger.d(
+        'Navigator didReplace ${routeToString(oldRoute)} → ${routeToString(newRoute)}');
+  }
+
+  String routeToString(Route<dynamic> route) {
+    return route?.settings?.name;
+  }
```
Comment (MarcelGarus) [inline lib/app/services/navigator_observer.dart]:
```suggestion
  String routeToString(Route<dynamic> route) => route?.settings?.name;
```

File: lib/app/services/network.dart [hunk 3]
```diff
@@ -2,11 +2,13 @@ import 'dart:convert';
 import 'dart:io';
 
 import 'package:flutter/widgets.dart';
-import 'package:get_it/get_it.dart';
 import 'package:http/http.dart' as http;
 import 'package:meta/meta.dart';
+import 'package:get_it/get_it.dart';
```
Comment (MarcelGarus) [inline lib/app/services/network.dart]: I'm not saying I know the alphabet very well, but this seems unordered…

**Review by MarcelGarus — CHANGES_REQUESTED** [review]

**Review by MarcelGarus — APPROVED** [review]

**Pull request merged**